File: dixiyang-engine/src/main/java/com/dixiyang/server/Service/chat/agent/PromptTemplates.java
package com.dixiyang.server.Service.chat.agent;

import dev.langchain4j.model.input.PromptTemplate;

import java.util.Map;

/**
 * 所有 Prompt 模板的统一管理
 * 使用 LangChain4j PromptTemplate，消除手写字符串拼接
 */
public final class PromptTemplates {

    private PromptTemplates() {}

    // ==================== System Prompt 模板 ====================

    private static final String BASE_RULES = """
        你是小说创作助手。严格遵循以下规则：
        1. 角色信息必须与【固定设定】中的角色卡完全一致
        2. 故事事件必须与【固定设定】中的故事节点一致
        3. 设定中没有的信息不能凭空编造，必须明确说"设定中没有这个信息"
        4. 如果用户要求写与设定矛盾的内容，拒绝并说明原因
        """;

    public static final PromptTemplate SYSTEM_PROMPT = PromptTemplate.from(BASE_RULES + """

        【对话模式：{{mode}}】
        {{modeInstruction}}
        {{customInstruction}}
        """);

    // ==================== 模式指令 ====================

    public static final Map<ConversationMode, String> MODE_INSTRUCTIONS = Map.of(
        ConversationMode.WRITE, "按设定创作剧情、对话、描写、续写。严格基于固定设定，不自由发挥。",
        ConversationMode.DISCUSS, "只回答关于角色、世界观、大纲的设定问题。不要创作内容。如果用户要求创作，提醒他切换到创作模式。",
        ConversationMode.ANALYZE, "分析角色性格、剧情逻辑、世界观一致性。给出有依据的分析。请输出 <thinking> 分析过程，再给结论。",
        ConversationMode.BRAINSTORM, "帮用户发散思考、寻找创作灵感。可以提出多种可能性。请输出 <thinking> 思考过程，再给建议。",
        ConversationMode.ASK, "快速精准回答用户问题，简短不啰嗦。"
    );

    // ==================== User Prompt 模板 ====================

    public static final PromptTemplate USER_PROMPT = PromptTemplate.from("""
        {{fixedContext}}
        {{history}}
        {{edits}}

        【用户本轮输入】
        {{userMessage}}
        """);

    // ==================== 修正要点提取模板 ====================

    public static final PromptTemplate KEYPOINT_EXTRACT = PromptTemplate.from("""
        你是修正要点提取器。对比以下 AI 原始输出和用户修正后的内容，提取修正要点。

        AI 原始输出：{{original}}

        用户修正后：{{edited}}

        请返回 JSON（不要 markdown 包裹）：
        {"keyPoint":"一句话总结修正要点（30字以内）","errorType":"错误类型"}

        错误类型只能是以下之一：
        - FACT_ERROR（事实/数据错误）
        - SETTING_VIOLATION（违背设定）
        - TONE_WRONG（语气/风格不对）
        - TOO_LONG（太啰嗦）
        - OFF_TOPIC（答非所问）
        - OTHER（其他）

        只返回 JSON，不要任何解释。
        """);

    // ==================== 标题生成模板 ====================

    public static final PromptTemplate TITLE_GENERATE = PromptTemplate.from("""
        你是一个对话标题生成器。请用 3-8 个字概括以下对话的主题，直接返回标题，不要任何解释、标点或引号。

        对话：
        {{conversation}}
        """);

    // ==================== 历史摘要模板 ====================

    public static final PromptTemplate HISTORY_SUMMARIZE = PromptTemplate.from("""
        请用 3-5 句话概括以下对话的核心内容（设定讨论、创作内容、关键决定）。直接返回摘要，不要解释。

        {{previousSummary}}

        新对话内容：
        {{conversation}}
        """);

    // ==================== 便捷方法 ====================

    /**
     * 构建 System Prompt
     */
    public static String buildSystemPrompt(ConversationMode mode, String customInstruction) {
        return SYSTEM_PROMPT.apply(Map.of(
            "mode", mode.label,
            "modeInstruction", MODE_INSTRUCTIONS.get(mode),
            "customInstruction", customInstruction != null && !customInstruction.isBlank()
                ? "\n【用户自定义指令】\n" + customInstruction : ""
        )).text();
    }

    /**
     * 构建 User Prompt
     */
    public static String buildUserPrompt(String fixedContext, String history, String edits, String userMessage) {
        return USER_PROMPT.apply(Map.of(
            "fixedContext", fixedContext.isEmpty() ? "" : "【固定设定·不可违背】\n" + fixedContext + "\n\n",
            "history", history.isEmpty() ? "" : "【对话历史】\n" + history + "\n\n",
            "edits", edits.isBlank() ? "" : edits + "\n\n",
            "userMessage", userMessage
        )).text();
    }

    /**
     * 构建修正要点提取 Prompt
     */
    public static String buildKeyPointPrompt(String original, String edited) {
        return KEYPOINT_EXTRACT.apply(Map.of(
            "original", original.length() > 500 ? original.substring(0, 500) : original,
            "edited", edited.length() > 500 ? edited.substring(0, 500) : edited
        )).text();
    }
}

File: DixyangFast/src/dixiyang/services/prompt_templates.py
"""
所有 Prompt 模板的统一管理
对齐 Spring 版 PromptTemplates.java
"""

from .conversation_mode import ConversationMode, MODE_META

# ==================== System Prompt ====================

_BASE_RULES = """你是小说创作助手。严格遵循以下规则：
1. 角色信息必须与【固定设定】中的角色卡完全一致
2. 故事事件必须与【固定设定】中的故事节点一致
3. 设定中没有的信息不能凭空编造，必须明确说"设定中没有这个信息"
4. 如果用户要求写与设定矛盾的内容，拒绝并说明原因"""


def build_system_prompt(mode: ConversationMode, custom_instruction: str | None = None) -> str:
    meta = MODE_META[mode]
    parts = [
        _BASE_RULES,
        "",
        f"【对话模式：{meta['label']}】",
        meta["instruction"],
    ]
    if custom_instruction:
        parts.append("")
        parts.append("【用户自定义指令】")
        parts.append(custom_instruction)
    return "\n".join(parts)


# ==================== User Prompt ====================

def build_user_prompt(
    fixed_context: str,
    history: str,
    edits: str,
    user_message: str,
) -> str:
    parts: list[str] = []
    if fixed_context:
        parts.append(f"【固定设定·不可违背】\n{fixed_context}\n")
    if history:
        parts.append(f"【对话历史】\n{history}\n")
    if edits:
        parts.append(f"{edits}\n")
    parts.append(f"【用户本轮输入】\n{user_message}")
    return "\n".join(parts)


# ==================== 修正要点提取 ====================

_KEYPOINT_TEMPLATE = """你是修正要点提取器。对比以下 AI 原始输出和用户修正后的内容，提取修正要点。

AI 原始输出：{original}

用户修正后：{edited}

请返回 JSON（不要 markdown 包裹）：
{{"keyPoint":"一句话总结修正要点（30字以内）","errorType":"错误类型"}}

错误类型只能是以下之一：
- FACT_ERROR（事实/数据错误）
- SETTING_VIOLATION（违背设定）
- TONE_WRONG（语气/风格不对）
- TOO_LONG（太啰嗦）
- OFF_TOPIC（答非所问）
- OTHER（其他）

只返回 JSON，不要任何解释。"""


def build_keypoint_prompt(original: str, edited: str) -> str:
    return _KEYPOINT_TEMPLATE.format(
        original=original[:500],
        edited=edited[:500],
    )


# ==================== 标题生成 ====================

_TITLE_TEMPLATE = """你是一个对话标题生成器。请用 3-8 个字概括以下对话的主题，直接返回标题，不要任何解释、标点或引号。

对话：
{conversation}"""


def build_title_prompt(conversation: str) -> str:
    return _TITLE_TEMPLATE.format(conversation=conversation)


# ==================== 历史摘要 ====================

_SUMMARIZE_TEMPLATE = """请用 3-5 句话概括以下对话的核心内容（设定讨论、创作内容、关键决定）。直接返回摘要，不要解释。

{previous_summary}

新对话内容：
{conversation}"""


def build_summarize_prompt(previous_summary: str, conversation: str) -> str:
    prev = f"已有摘要：\n{previous_summary}" if previous_summary else ""
    return _SUMMARIZE_TEMPLATE.format(
        previous_summary=prev,
        conversation=conversation,
    )


# ==================== 角色设定提取 ====================

_CHARACTER_EXTRACT_TEMPLATE = """你是一个角色设定提取器。分析以下对话内容，提取与角色相关的结构化信息。

对话内容：
{conversation}

请返回 JSON 格式的角色设定（不要 markdown 包裹）：
{{
  "势力": "角色所属势力/组织",
  "技能": ["技能1", "技能2"],
  "武器": "武器名称",
  "弱点": ["弱点1", "弱点2"],
  "性格特点": "性格描述",
  "外貌特征": "外貌描述",
  "背景故事": "背景描述",
  "人际关系": {{
    "角色名": "关系描述"
  }},
  "其他信息": "未归类的重要信息"
}}

规则：
1. 只返回对话中明确提到的信息，不要编造
2. 如果某个字段没有相关信息，不要包含该字段
3. 只返回 JSON，不要任何解释
4. 如果对话中没有任何可提取的设定信息，返回空对象 {{}}"""


def build_character_extract_prompt(conversation: str) -> str:
    return _CHARACTER_EXTRACT_TEMPLATE.format(conversation=conversation)
